File: build-support/java/bootstrap.py
#!/usr/bin/env python3
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy at http://www.apache.org/licenses/LICENSE-2.0
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Checksum-pinned archive bootstrapper for Aurora's in-place Java build."""
import hashlib
import os
from pathlib import Path, PurePosixPath
import shutil
import stat
import tarfile
import urllib.request
import zipfile

class BuildError(Exception):
    pass

def require(condition, message):
    if not condition:
        raise BuildError(message)

def sha256(path):
    digest = hashlib.sha256()
    with path.open('rb') as stream:
        for block in iter(lambda: stream.read(1024 * 1024), b''):
            digest.update(block)
    return digest.hexdigest()

def real_path(path):
    require(path.is_absolute(), 'Path must be absolute: ' + str(path))
    require(not any(item.is_symlink() for item in (path, *path.parents)),
            'Symlink path rejected: ' + str(path))

def archive_member(root, name):
    value = PurePosixPath(name)
    require(name and not value.is_absolute() and '..' not in value.parts and '\\' not in name,
            'Unsafe archive entry: ' + name)
    target = root.joinpath(*value.parts)
    real_path(target)
    return target

def extract(archive, destination):
    """Extract regular files while rejecting traversal and special files."""
    require(not destination.exists(), 'Extraction destination must be new')
    destination.mkdir(mode=0o700)
    seen, total = set(), 0
    def reserve(name, size):
        nonlocal total
        target = archive_member(destination, name)
        require(target not in seen, 'Duplicate archive entry')
        seen.add(target)
        require(0 <= size <= 512 * 1024 * 1024 and total + size <= 2 * 1024 ** 3,
                'Tool archive is too large')
        total += size
        return target
    if archive.suffix == '.zip':
        with zipfile.ZipFile(archive) as source:
            for item in source.infolist():
                target = reserve(item.filename, item.file_size)
                mode = item.external_attr >> 16
                require(not stat.S_ISLNK(mode), 'ZIP symlink rejected')
                require(stat.S_IFMT(mode) in (0, stat.S_IFREG, stat.S_IFDIR), 'ZIP special file')
                if item.is_dir():
                    target.mkdir(parents=True, exist_ok=True)
                else:
                    target.parent.mkdir(parents=True, exist_ok=True)
                    with source.open(item) as src, target.open('xb') as dst:
                        shutil.copyfileobj(src, dst)
                    target.chmod(0o755 if mode & 0o111 else 0o644)
        return
    links = []
    with tarfile.open(archive, 'r:*') as source:
        for item in source:
            target = reserve(item.name, item.size)
            if item.isdir():
                target.mkdir(parents=True, exist_ok=True)
            elif item.isfile():
                target.parent.mkdir(parents=True, exist_ok=True)
                source_file = source.extractfile(item)
                require(source_file is not None, 'Unreadable archive member')
                with source_file, target.open('xb') as dst:
                    shutil.copyfileobj(source_file, dst)
                target.chmod(0o755 if item.mode & 0o111 else 0o644)
            elif item.issym() or item.islnk():
                links.append((target, item.linkname, item.islnk()))
            else:
                raise BuildError('Archive special file rejected')
    for target, name, hard in links:
        require(not Path(name).is_absolute() and '\\' not in name, 'Absolute archive link')
        source = ((destination if hard else target.parent) / name).resolve()
        require(source.is_relative_to(destination) and
                (source.is_file() or source.is_dir() and not hard),
                'Escaping/dangling archive link')
        target.parent.mkdir(parents=True, exist_ok=True)
        if source.is_dir():
            require(not target.is_relative_to(source), 'Cyclic archive directory link')
            shutil.copytree(source, target)
        else:
            shutil.copyfile(source, target)
            shutil.copymode(source, target)

def get_tools(pins, cache, work, seed=None, offline=False):
    archives = cache / 'archives'
    archives.mkdir(parents=True, exist_ok=True, mode=0o700)
    result = {}
    for name, pin in pins['tools'].items():
        archive = archives / pin['file']
        real_path(archive)
        if archive.exists():
            require(archive.is_file() and sha256(archive) == pin['sha256'],
                    'Cached tool checksum mismatch: ' + name)
        else:
            temporary = archive.with_name(archive.name + '.part')
            real_path(temporary)
            require(not temporary.exists(), 'Incomplete download exists: ' + str(temporary))
            try:
                candidate = seed / pin['file'] if seed is not None else None
                if candidate is not None and candidate.is_file():
                    require(sha256(candidate) == pin['sha256'], 'Seed tool checksum mismatch: ' + name)
                    shutil.copyfile(candidate, temporary)
                else:
                    require(not offline, 'Offline tool missing: ' + name)
                    require(pin['url'].startswith('https://'), 'Tools require HTTPS')
                    print('Downloading pinned ' + name, flush=True)
                    with urllib.request.urlopen(pin['url'], timeout=60) as source, temporary.open('xb') as target:
                        require(source.url.startswith('https://'), 'Insecure tool redirect')
                        shutil.copyfileobj(source, target)
                require(sha256(temporary) == pin['sha256'], 'Tool checksum mismatch: ' + name)
                os.replace(temporary, archive)
            finally:
                if temporary.is_file():
                    temporary.unlink()
        destination = work / name
        extract(archive, destination)
        result[name] = destination / pin['directory']
        require(result[name].is_dir(), 'Tool archive layout differs: ' + name)
    return result

File: tools/internal/bootstrap/bootstrap.go
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy at http://www.apache.org/licenses/LICENSE-2.0
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package bootstrap

import (
	"context"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"syscall"
	"time"
)

const MaxArchiveBytes int64 = 1 << 30

type Pin struct {
	File      string `json:"file"`
	URL       string `json:"url"`
	SHA256    string `json:"sha256"`
	Directory string `json:"directory"`
	Version   string `json:"version,omitempty"`
}
type Manifest struct {
	Schema   int            `json:"schema"`
	Platform string         `json:"platform"`
	Tools    map[string]Pin `json:"tools"`
}

func ReadJSON(path string, value any) error {
	f, err := os.Open(path)
	if err != nil {
		return err
	}
	defer f.Close()
	data, err := io.ReadAll(io.LimitReader(f, 1<<20+1))
	if err != nil {
		return err
	}
	if len(data) > 1<<20 {
		return errors.New("manifest size limit")
	}
	// Reject duplicate keys even in metadata fields that this command does not use.
	decoder := json.NewDecoder(strings.NewReader(string(data)))
	if err := uniqueJSON(decoder); err != nil {
		return err
	}
	if _, err := decoder.Token(); err != io.EOF {
		return errors.New("trailing manifest data")
	}
	return json.Unmarshal(data, value)
}
func uniqueJSON(d *json.Decoder) error {
	t, err := d.Token()
	if err != nil {
		return err
	}
	if delim, ok := t.(json.Delim); ok {
		switch delim {
		case '{':
			seen := map[string]bool{}
			for d.More() {
				key, err := d.Token()
				if err != nil {
					return err
				}
				s, ok := key.(string)
				if !ok || seen[s] {
					return errors.New("duplicate JSON key")
				}
				seen[s] = true
				if err := uniqueJSON(d); err != nil {
					return err
				}
			}
		case '[':
			for d.More() {
				if err := uniqueJSON(d); err != nil {
					return err
				}
			}
		default:
			return errors.New("invalid JSON delimiter")
		}
		_, err = d.Token()
		return err
	}
	return nil
}
func LoadManifest(path string) (Manifest, error) {
	var m Manifest
	if err := ReadJSON(path, &m); err != nil {
		return m, err
	}
	if m.Schema != 1 || m.Platform != runtime.GOOS+"/"+runtime.GOARCH || len(m.Tools) == 0 {
		return m, errors.New("unsupported tool manifest schema or host platform")
	}
	for name, pin := range m.Tools {
		if !safeComponent(name) {
			return m, errors.New("unsafe tool name")
		}
		if err := pin.Validate(); err != nil {
			return m, fmt.Errorf("%s: %w", name, err)
		}
	}
	return m, nil
}
func safeComponent(value string) bool {
	return value != "" && value != "." && value != ".." && !strings.ContainsAny(value, "/\\")
}
func (p Pin) Validate() error {
	if !safeComponent(p.File) {
		return errors.New("unsafe archive filename")
	}
	if len(p.SHA256) != 64 || strings.ToLower(p.SHA256) != p.SHA256 {
		return errors.New("invalid SHA256 pin")
	}
	if _, err := hex.DecodeString(p.SHA256); err != nil {
		return err
	}
	u, err := url.Parse(p.URL)
	if err != nil || u.Scheme != "https" || u.Host == "" || u.User != nil {
		return errors.New("tool archives require HTTPS")
	}
	if p.Directory != "" {
		if _, err := MemberPath("/private", p.Directory); err != nil {
			return err
		}
	}
	return nil
}

// CheckPath examines components before cleaning, including a symlink followed
// by /..; normalization must not hide a symlink the caller supplied.
func CheckPath(path string) (string, error) {
	if !filepath.IsAbs(path) {
		cwd, err := os.Getwd()
		if err != nil {
			return "", err
		}
		path = cwd + "/" + path
	}
	current := "/"
	for _, part := range strings.Split(path, "/") {
		if part == "" || part == "." {
			continue
		}
		current = filepath.Join(current, part)
		info, err := os.Lstat(current)
		if err != nil {
			if os.IsNotExist(err) {
				continue
			}
			return "", err
		}
		if info.Mode()&os.ModeSymlink != 0 {
			return "", fmt.Errorf("symlink path rejected: %s", current)
		}
	}
	return filepath.Clean(path), nil
}
func PrivateCache(path string) (string, error) {
	clean, err := CheckPath(path)
	if err != nil {
		return "", err
	}
	if err = os.MkdirAll(clean, 0700); err != nil {
		return "", err
	}
	info, err := os.Stat(clean)
	if err != nil {
		return "", err
	}
	owner, ok := info.Sys().(*syscall.Stat_t)
	if !ok || int(owner.Uid) != os.Getuid() || !info.IsDir() {
		return "", errors.New("cache must be a directory owned by this user")
	}
	if err = os.Chmod(clean, 0700); err != nil {
		return "", err
	}
	return clean, nil
}
func Lock(path string) (func(), error) {
	clean, err := CheckPath(path)
	if err != nil {
		return nil, err
	}
	f, err := os.OpenFile(clean, os.O_CREATE|os.O_RDWR|syscall.O_NOFOLLOW|syscall.O_NONBLOCK, 0600)
	if err != nil {
		return nil, err
	}
	info, err := f.Stat()
	if err != nil || !info.Mode().IsRegular() {
		f.Close()
		return nil, errors.New("lock must be a regular file")
	}
	if err = syscall.Flock(int(f.Fd()), syscall.LOCK_EX); err != nil {
		f.Close()
		return nil, err
	}
	return func() { syscall.Flock(int(f.Fd()), syscall.LOCK_UN); f.Close() }, nil
}
func Digest(path string) (string, error) {
	clean, err := CheckPath(path)
	if err != nil {
		return "", err
	}
	f, err := os.OpenFile(clean, os.O_RDONLY|syscall.O_NOFOLLOW|syscall.O_NONBLOCK, 0)
	if err != nil {
		return "", err
	}
	defer f.Close()
	info, err := f.Stat()
	if err != nil || !info.Mode().IsRegular() {
		return "", errors.New("digest requires regular file")
	}
	h := sha256.New()
	if _, err = io.Copy(h, f); err != nil {
		return "", err
	}
	return hex.EncodeToString(h.Sum(nil)), nil
}
func verify(path, digest string) error {
	info, err := os.Lstat(path)
	if err != nil {
		return err
	}
	if !info.Mode().IsRegular() || info.Size() > MaxArchiveBytes {
		return errors.New("archive type or size limit")
	}
	actual, err := Digest(path)
	if err != nil {
		return err
	}
	if actual != digest {
		return fmt.Errorf("archive checksum mismatch: %s", path)
	}
	return nil
}

func EnsureArchive(ctx context.Context, pin Pin, cache, seed string, offline bool) (string, error) {
	if err := pin.Validate(); err != nil {
		return "", err
	}
	archives, err := PrivateCache(filepath.Join(cache, "archives"))
	if err != nil {
		return "", err
	}
	path, err := CheckPath(filepath.Join(archives, pin.File))
	if err != nil {
		return "", err
	}
	if _, err = os.Lstat(path); err == nil {
		return path, verify(path, pin.SHA256)
	} else if !os.IsNotExist(err) {
		return "", err
	}
	temporary := path + ".part"
	if _, err = CheckPath(temporary); err != nil {
		return "", err
	}
	target, err := os.OpenFile(temporary, os.O_WRONLY|os.O_CREATE|os.O_EXCL|syscall.O_NOFOLLOW, 0600)
	if err != nil {
		return "", err
	}
	defer os.Remove(temporary)
	defer target.Close()
	var input io.ReadCloser
	if seed != "" {
		candidate, err := CheckPath(filepath.Join(seed, pin.File))
		if err != nil {
			return "", err
		}
		if _, err = os.Lstat(candidate); err == nil {
			if err = verify(candidate, pin.SHA256); err != nil {
				return "", err
			}
			input, err = os.OpenFile(candidate, os.O_RDONLY|syscall.O_NOFOLLOW|syscall.O_NONBLOCK, 0)
			if err != nil {
				return "", err
			}
		} else if !os.IsNotExist(err) {
			return "", err
		}
	}
	if input == nil {
		if offline {
			return "", fmt.Errorf("offline tool archive missing: %s", pin.File)
		}
		request, err := http.NewRequestWithContext(ctx, http.MethodGet, pin.URL, nil)
		if err != nil {
			return "", err
		}
		client := &http.Client{Timeout: 60 * time.Second, CheckRedirect: func(r *http.Request, via []*http.Request) error {
			if r.URL.Scheme != "https" || len(via) > 10 {
				return errors.New("insecure/excessive tool redirect")
			}
			return nil
		}}
		response, err := client.Do(request)
		if err != nil {
			return "", err
		}
		if response.StatusCode != 200 {
			response.Body.Close()
			return "", fmt.Errorf("tool download HTTP %d", response.StatusCode)
		}
		input = response.Body
	}
	defer input.Close()
	copied, err := io.Copy(target, io.LimitReader(input, MaxArchiveBytes+1))
	if err != nil {
		return "", err
	}
	if copied > MaxArchiveBytes {
		return "", errors.New("tool download size limit")
	}
	if err = target.Sync(); err != nil {
		return "", err
	}
	if err = target.Close(); err != nil {
		return "", err
	}
	if err = verify(temporary, pin.SHA256); err != nil {
		return "", err
	}
	if err = os.Link(temporary, path); err != nil {
		return "", err
	}
	return path, nil
}
func Tools(ctx context.Context, m Manifest, cache, work, seed string, offline bool) (map[string]string, error) {
	result := map[string]string{}
	for name, pin := range m.Tools {
		archive, err := EnsureArchive(ctx, pin, cache, seed, offline)
		if err != nil {
			return nil, err
		}
		destination := filepath.Join(work, name)
		if err = Extract(ctx, archive, destination, false); err != nil {
			return nil, err
		}
		tool := filepath.Join(destination, pin.Directory)
		info, err := os.Stat(tool)
		if err != nil || !info.IsDir() {
			return nil, fmt.Errorf("tool archive layout differs: %s", name)
		}
		result[name] = tool
	}
	return result, nil
}
